TrkDiag/src/TrackMatching_module.cc

```diff
@@ -0,0 +1,235 @@
+
+    // Retrieve the hit lists
+    const auto& hits_1 = k_1->hits();
+    const auto& hits_2 = k_2->hits();
+    if(hits_1.empty() || hits_2.empty()) return false;
+
+    // Count the number of overlapping hits
+    int overlap(0);
+    for(auto hit_1 : hits_1) {
+      for(auto hit_2 : hits_2) {
+        if(_useIndices) { // match using the digi index
+          if(hit_1._index == hit_2._index) ++overlap;
+        } else {
+          if(hit_1._sid == hit_2._sid && std::fabs(hit_1.hitTime() - hit_2.hitTime()) < _timeWindow) ++overlap;
+        }
+      }
+    }
+    const double fraction = overlap * 2. / (hits_1.size() + hits_2.size());
+    const bool pass = fraction >= _threshold;
+    if(_debug > 2) printf("  Track overlap: %2i hits, %5.3f fraction --> pass = %o\n", overlap, fraction, pass);
+    if(_makeHists) {
+      _hists.overlap->Fill(fraction);
+    }
+    return pass;
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  // Merge two track clusters into the first cluster
+  void TrackMatching::merge_clusters(KalSeedCluster& cluster, KalSeedCluster& cluster_j) {
+    for(const auto& ptr : cluster_j) cluster.push_back(ptr); // merge into the first cluster
+    cluster_j = {}; // empty the second cluster
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  // Create track clusters from input track collections
+  void TrackMatching::create_clusters(std::vector<KalSeedCluster>& clusters) {
+
+    const size_t ntrks = clusters.size();
+    if(_debug > 2) printf("[TrackMatching::%s::%s] Inspecting %zu tracks\n",
+                          __func__, moduleDescription().moduleLabel().c_str(), ntrks);
+
+    // Loop through the clusters, merging overlapping clusters
+    for(size_t index = 0; index < ntrks; ++index) {
+      KalSeedCluster& cluster = clusters[index];
+      if(cluster.empty()) continue; // already merged into an earlier cluster
+      if(cluster.size() > 1) // input clusters should always be 1 track, as matches are clustered into lower index clusters
+        throw cet::exception("RECO") << "mu2e::TrackMatching::" << __func__ << ": Input track cluster has already been clustered! Size = " << cluster.size();
+      if(_debug > 2) printf("  Checking track cluster %zu for overlaps\n", index);
+
+      // Check each following cluster (of size 1) for overlapping tracks
+      for(size_t jtrk = index + 1; jtrk < ntrks; ++jtrk) {
+        KalSeedCluster& cluster_j = clusters[jtrk];
+        if(cluster_j.empty()) continue;
+        if(cluster_j.size() > 1) // input clusters should always be 1 track, as matches are clustered into lower index clusters
+          throw cet::exception("RECO") << "mu2e::TrackMatching::" << __func__ << ": Track cluster matching against has already been clustered! Size = " << cluster_j.size();
+
+        // Get the tracks to compare
+        KalSeedPtr& k_i = cluster.front();
```

Comment:
You test only against the 1st KalSeed in the cluster. An alternative is to test against all current cluster members and merge if any match. That's a more aggressive algorithm and in some cases can lead to over-clustering, but I think it would be safe in your case. Neither algorithm is robust against changes in input order: that would require testing all possible pairings and merging the 'best' matches first. That's more complicated to implement and is probably overkill for our case. Just options to consider.

Reply:
Each track cluster on input has size 1, and every later cluster only have 1 track so I was thinking it's always 1 size comparisons. I forgot that tracks that get clustered then don't get this comparison loop done as they've been clustered and moved. I've fixed this oversight, where I now keep moving through the track cluster performing the matching as tracks are added to the cluster until all have been checked, then I move to the next non-clustered track.